src/lndk_offers.rs

```diff
@@ -0,0 +1,237 @@
+pub(crate) enum OfferError<Secp256k1Error> {
+    BuildUIRFailure(Bolt12SemanticError),
+    SignFailure(SignError<Secp256k1Error>),
+}
+
+impl Display for OfferError<Secp256k1Error> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            OfferError::BuildUIRFailure(e) => write!(f, "{:?}", e),
+            OfferError::SignFailure(e) => write!(f, "{:?}", e),
+        }
+    }
+}
+
+impl Error for OfferError<Secp256k1Error> {}
+
+// Decodes a bech32 string into an LDK offer.
+pub fn decode(offer_str: String) -> Result<Offer, Bolt12ParseError> {
+    offer_str.parse::<Offer>()
+}
+
+#[allow(dead_code)]
+// create_request_invoice builds and signs an invoice request, the first step in the BOLT 12 process of paying an offer.
+pub(crate) async fn create_request_invoice(
+    mut signer: impl MessageSigner + std::marker::Send + 'static,
+    offer: Offer,
+    metadata: Vec<u8>,
+    network: Network,
+    msats: u64,
+) -> Result<InvoiceRequest, OfferError<bitcoin::secp256k1::Error>> {
+    let key_loc = KeyLocator {
+        key_family: 6,
+        key_index: 1,
+    };
+
+    let pubkey_bytes = signer
+        .derive_key(key_loc.clone())
+        .await
+        .expect("failed to get key");
+    let pubkey = PublicKey::from_slice(&pubkey_bytes).expect("failed to deserialize public key");
+
+    let unsigned_invoice_req = offer
+        .request_invoice(metadata, pubkey)
+        .unwrap()
+        .chain(network)
+        .unwrap()
+        .amount_msats(msats)
+        .unwrap()
+        .build()
+        .map_err(OfferError::BuildUIRFailure)?;
+
+    // To create a valid invoice request, we also need to sign it. This is spawned in a blocking
+    // task because we need to call block_on on sign_message so that sign_closure can be a
+    // synchronous closure.
+    task::spawn_blocking(move || {
+        let sign_closure = |msg: &UnsignedInvoiceRequest| {
+            let tagged_hash = msg.as_ref();
+            let tag = tagged_hash.tag().to_string();
+
+            let signature = block_on(signer.sign_message(key_loc, tagged_hash.merkle_root(), tag))
+                .map_err(|_| Secp256k1Error::InvalidSignature)?;
+
+            Signature::from_slice(&signature)
+        };
+
+        unsigned_invoice_req
+            .sign(sign_closure)
+            .map_err(OfferError::SignFailure)
+    })
+    .await
+    .unwrap()
+}
+
+/// MessageSigner provides a layer of abstraction over the LND API for message signing.
+#[async_trait]
+pub(crate) trait MessageSigner {
+    async fn derive_key(&mut self, key_loc: KeyLocator) -> Result<Vec<u8>, Status>;
+    async fn sign_message(
+        &mut self,
+        key_loc: KeyLocator,
+        merkle_hash: Hash,
+        tag: String,
+    ) -> Result<Vec<u8>, Status>;
+}
+
+/// Bolt12Signer is responsible for signing the InvoiceRequest.
+#[derive(Clone)]
+pub(crate) struct Bolt12Signer {
+    client: Client,
+}
```

Comment:
Do we need this extra layer of abstraction? Couldn't we implement `MessageSigner` on `BlockingClient` directly? Since this is essentially just a straight cut-through to the underlying `Client`.

Reply:
Oh epic. For some reason I thought in rust you couldn't add methods to a struct defined elsewhere but looks like that's wrong. *insert Eric Wareheim mind blown gif here*